// typedoc-plugin-markdown/dist/resources/helpers/comment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Handlebars = require("handlebars");
const URL_PREFIX = /^(http|ftp)s?:\/\//;
const BRACKETS = /\[\[([^\]]+)\]\]/g;
const INLINE_TAG = /(?:\[(.+?)\])?\{@(link|linkcode|linkplain)\s+((?:.|\n)+?)\}/gi;
function default_1(theme) {
    Handlebars.registerHelper('comment', function () {
        const { project } = theme;
        function replaceBrackets(text) {
            return text.replace(BRACKETS, (match, content) => {
                const split = splitLinkText(content);
                return buildLink(match, split.target, split.caption);
            });
        }
        function replaceInlineTags(text) {
            return text.replace(INLINE_TAG, (match, leading, tagName, content) => {
                const split = splitLinkText(content);
                const target = split.target;
                const caption = leading || split.caption;
                return buildLink(match, target, caption, tagName === 'linkcode');
            });
        }
        function buildLink(original, target, caption, monospace = false) {
            if (monospace) {
                caption = '`' + caption + '`';
            }
            if (URL_PREFIX.test(target)) {
                return `[${caption}](${target})`;
            }
            const reflection = project === null || project === void 0 ? void 0 : project.findReflectionByName(target);
            if (reflection && reflection.url) {
                return `[${caption}](${Handlebars.helpers.relativeURL(reflection.url)})`;
            }
            else {
                return original;
            }
        }
        function splitLinkText(text) {
            let splitIndex = text.indexOf('|');
            if (splitIndex === -1) {
                splitIndex = text.search(/\s/);
            }
            if (splitIndex !== -1) {
                return {
                    caption: text
                        .substr(splitIndex + 1)
                        .replace(/\n+/, ' ')
                        .trim(),
                    target: text.substr(0, splitIndex).trim(),
                };
            }
            else {
                return {
                    caption: text,
                    target: text,
                };
            }
        }
        return replaceInlineTags(replaceBrackets(this));
    });
}
exports.default = default_1;
